File: scripts/revision_probe.py
# SPDX-License-Identifier: Apache-2.0
import json
import sys
from psp_cdl_mcpproxy.mcp import HttpMcpClient, StdioMcpClient
from psp_cdl_mcpproxy import McpDispatchGate
from revision_fixtures import APPROVAL
from dispatch_fixtures import Fixture
c=json.loads(sys.argv[1]);p=n=f=None
def connect(config): return HttpMcpClient.connect(config,lambda _:"test-downstream") if c["transport"]=="http" else StdioMcpClient.connect(config)
try:
    p=connect(c["config"]);f=Fixture()
    regs=p.registrations("echo",[APPROVAL],f.now,"wrong" if c.get("unapproved") else p.catalog_snapshot["approvalDigest"])
    gate=McpDispatchGate(f.store,f,"registry-1",regs)
    def call(): return gate.call_tool("test-owner",f.session["sessionId"],{"name":"echo.read","arguments":{"message":"hello 🧪"}},f.options)
    result=call()
    if c.get("nextConfig"):
        n=connect(c["nextConfig"])
        candidate=n.registrations("echo",[{**APPROVAL,"revision":"tool-2"}],f.now,n.catalog_snapshot["approvalDigest"])
        gate.replace_registry("registry-1","registry-2",candidate);f.flags["registryDrift"]=True;p.close()
        result=call()
    print(json.dumps({"code":"OK","released":1,"data":result["data"],"toolRevision":result["provenance"]["toolRevision"],"registryRevision":result["provenance"]["registryRevision"]},ensure_ascii=False))
except Exception as exc: print(json.dumps({"code":getattr(exc,"code","UNEXPECTED_ERROR"),"released":0}))
finally:
    if p: p.close()
    if n: n.close()
    if f: f.close()

File: scripts/revision-probe.mjs
// SPDX-License-Identifier: Apache-2.0
import {HttpMcpClient,StdioMcpClient} from '@psp-cdl/mcpproxy/mcp';
import {McpDispatchGate} from '@psp-cdl/mcpproxy';
import {fixture} from './dispatch-fixtures.mjs';
import {approval} from './revision-fixtures.mjs';
const c=JSON.parse(process.argv[2]);let p,n,f;
const connect=cfg=>c.transport==='http'?HttpMcpClient.connect(cfg,()=> 'test-downstream'):StdioMcpClient.connect(cfg);
try {
  p=await connect(c.config);f=await fixture();
  const regs=p.registrations('echo',[approval],()=>1000,c.unapproved?'wrong':p.catalogSnapshot.approvalDigest);
  const gate=new McpDispatchGate(f.store,f.host,'registry-1',regs),call=()=>gate.callTool('test-owner',f.session.sessionId,{name:'echo.read',arguments:{message:'hello 🧪'}},f.options);
  let result=await call();
  if(c.nextConfig){
    n=await connect(c.nextConfig);
    const next=n.registrations('echo',[{...approval,revision:'tool-2'}],()=>1000,n.catalogSnapshot.approvalDigest);
    gate.replaceRegistry('registry-1','registry-2',next);f.flags.registryDrift=true;await p.close();
    result=await call();
  }
  console.log(JSON.stringify({code:'OK',released:1,data:result.data,toolRevision:result.provenance.toolRevision,registryRevision:result.provenance.registryRevision}));
}catch(e){console.log(JSON.stringify({code:e.code??'UNEXPECTED_ERROR',released:0}));}
finally{await p?.close();await n?.close();f?.close();}
